fleetspeak/src/server/stats.go

```diff
@@ -59,6 +62,132 @@ func (s noopStatsCollector) ResourceUsageDataReceived(cd *db.ClientData, rud mpb
+		Help: "The number of backlog messages received from datastore by Fleetspeak server",
+	})
+
+	messagesReceivedFromClient = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_messages_ingested_messages_received_from_client",
+		Help: "The number of messages received from a client by Fleetspeak server",
+	})
+
+	messagesSaved = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_messages_saved_total",
+		Help: "The total number of messages saved by Fleetspeak server",
+	})
+
+	messagesSavedSize = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_messages_saved_payload_size",
+		Help: "The total payload size of messages saved by Fleetspeak server (in bytes)",
+	})
+
+	messagesProcessed = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_server_messages_processed_total",
+		Help: "The total number of messages processed by Fleetspeak server",
+	})
+
+	messagesErrored = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_server_messages_errored_total",
+		Help: "The total number of message processings that returned an error (temporary, or permanent)",
+	})
+
+	messagesErroredTemp = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_server_messages_errored_temporary",
+		Help: "The total number of message processings that returned an error (temporary)",
+	})
+
+	messagesErroredPerm = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_server_messages_errored_permanent",
+		Help: "The total number of message processings that returned an error (permanent)",
+	})
+
+	messagesDropped = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_server_messages_dropped_total",
+		Help: "The total number of messages dropped by Fleetspeak server when too many messages for the sevices are being processed.",
+	})
+
+	clientPolls = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_server_client_polls_total",
+		Help: "The total number of times a client polls the Fleetspeak server.",
+	})
+
+	datastoreOperationsCompleted = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_server_datastore_operations_completed_total",
+		Help: "The total number of datastore operations completed.",
+	})
+
+	resourcesUsageDataReceived = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_server_resource_usage_data_received_total",
+		Help: "The total number of times a client-resource-usage proto is received.",
+	})
+
+	killNotificationsReceived = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_server_kill_notifications_received_total",
+		Help: "The total number of times a kill notification is received from a client.",
+	})
+)
+
+// A prometheusStatsCollector is an implementation of a Collector interface.
+// It exports stats to a Prometheus HTTP handler, which are exposed at :2112/metrics
+// and are scrapable by Prometheus.
+type prometheusStatsCollector struct{}
```

Comment:
There is a lot of additional metrics that can be added here, especially as histograms. Obvious things: database operation time, message processing time, client memory usage, client cpu usage, and probably a lot more that I'm forgetting.

Also, most of these things can and should be broken down by one or more parameter. If you are running multiple services it is very helpful to know which service's message (or which type of which service's message) is erroring, not that messages are erroring. It is always helpful to know which database operations are slow, etc, etc.

Glancing at the prometheous godoc, I think you want the *Vec types for most of these.

Anyway, I think this is probably fine to pull as a starting point, but I don't think it is complete enough to close #191

Reply:
Regarding breaking down the more parameters - I am not sure how those would be gathered. For instance, in the `MessageSaved` [metric](https://github.com/google/fleetspeak/blob/7ccefefe98e1e92af7559fb1d849bb1eb2ea3c48/fleetspeak/src/server/stats.go#L153), a lot of data is gathered, like the service used, the message type and so on. These can vary depending on the system being used (which is not necessarily GRR), and so unlike the rather simple [Counters](https://github.com/google/fleetspeak/blob/7ccefefe98e1e92af7559fb1d849bb1eb2ea3c48/fleetspeak/src/server/stats.go#L83) I have implemented for it so far, these kind of metrics would have to somehow dynamically create new Counters/Gauges/etc.
Am I going about this the right way, or should it be simpler than that?

I'll make sure to add metrics in a different, dedicated PR so it's easier to review and maintain. I'll also link to this PR and discussion for reference.

Reply:
My understanding is that Prometheus labels is what has to be used here (see 
https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels and https://prometheus.io/docs/practices/naming/).

Labels effectively introduce additional dimensions to the metric. Therefore it's important that each label has a finite number of possible values. This assumption holds for thinks like service name (there's a finite number of services in a particular fleetspeak deployment), database operation type, etc.

For example, in case of [MessageSaved](https://github.com/google/fleetspeak/blob/298e9ae7573cf0c2f29b0c1e68a0a11370259c53/fleetspeak/src/server/stats/collector.go#L94), `service`, `messageType` and `forClient` should likely be labels. While `savedPayloadBytes` is what will get accumulated in the metric value.

In any case, figuring this out in a separate PT sounds good.

Reply:
I've never used Prometheus, but it is similar to what we use internally, and internally, we find it  critical to split metrics by various attributes.

Again, looking through the docs, the *Vec Prometheus types seem to be what we need here, look at the example code in: https://godoc.org/github.com/prometheus/client_golang/prometheus#CounterVec.With

Reply:
+1 to Ben's suggestion. Please see the example at https://godoc.org/github.com/prometheus/client_golang/prometheus#CounterVec:

```
httpReqs := prometheus.NewCounterVec(
    prometheus.CounterOpts{
        Name: "http_requests_total",
        Help: "How many HTTP requests processed, partitioned by status code and HTTP method.",
    },
    []string{"code", "method"},
)
prometheus.MustRegister(httpReqs)

httpReqs.WithLabelValues("404", "POST").Add(42)
```

Here `code` and `method` are Prometheus labels.